### DISCUSSION TITLE
Use Redis Cache with Orchard Core

### DISCUSSION THREAD

**Discussion by miguelisidoro** [OrchardCMS/OrchardCore#15920] [Q&A]
Hello,

I have a big solution that is using Orchard core as its backend and for now it is not running Distributed cache.
I would like to use Redis Cache as distributed cache. We are about to have 2 Azure App Service instances in our deployment (horizontal scaling) and a distributed cache system is necessary.

Is it possible to provide a full step by step guide on how to configure Redis Cache in a Orchard Core based solution?
We would like to configure all the settings necessary in Orchard Core by code if possible.

I have Googled, searched the Orchard Core official docs and searched the issues and discussions and couldn't find a detailed solution.
Can you help?

Thanks

**Comment by Piedone** [comment]
While docs on Redis are indeed sparse, the gist of it is this:

1. Configure the Redis settings, see [here](https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore.Cms.Web/appsettings.json#L102). This can be done in any of the [configuration options](https://docs.orchardcore.net/en/latest/reference/core/Configuration/), just the sample is in an appsettings file.
2. Enable all Redis features from the admin (or else). Though in Azure you don't necessarily need Redis DataProtection, since for that you can use the Blob Storage-using one too: https://docs.orchardcore.net/en/latest/reference/modules/DataProtection.Azure/.

**Reply by miguelisidoro** [reply]
@Piedone thanks for the reply. 

About 2), is it enough to enable all Redis features on the default tenant and all other tenants will inherit Redis settings from the default tenant?

Thanks

**Reply by Piedone** [reply]
You'll need to enable it on all tenants. You don't need to do that one by one though, but like this too from your web app's `Program`: 

```c#
builder.Services
    .AddOrchardCms()
    // Need to list all their dependencies too.
    .AddTenantFeatures("OrchardCore.Redis", "OrchardCore.Redis.Cache");
```

**Comment by miguelisidoro** [comment]
@Piedone thanks.

1 - When you say Program, do you mean the Startup?
2 - Any other feature we should enable besides OrchardCore.Redis.Cache?

I have in the admin the following features:

![image](https://github.com/OrchardCMS/OrchardCore/assets/5429152/f31f4aeb-26d9-4524-b18b-6b960f1867e9)

If yes, can you send the features names we should reference in the code?

Thanks

**Reply by Piedone** [reply]
1. Yes.
2. See the `Id`s [here](https://github.com/OrchardCMS/OrchardCore/blob/1d4589bb9f0b9dc30fcb452655757df540eaffd1/src/OrchardCore.Modules/OrchardCore.Redis/Manifest.cs).

**Reply by miguelisidoro** [reply]
@Piedone thanks.

By including the AddTenantFeatures in Startup, we guarantee that Redis features will be enabled on all tenants right?

Thanks

**Reply by Piedone** [reply]
Yes, and they can't be disabled from the UI.

**Reply by miguelisidoro** [reply]
@Piedone thanks.

Just tried and I confirm that features are enabled in my development tenant and can't be disabled from the UI.

A few doubts on what happens next...

1 - How can we check that everything is OK after configuring Redis settings and activating Redis features? What should I be looking into? The goal is to check that Redis is being used in my development tenant.
2 - Currently we have in development a json file in the DEVBOX file system with the list of tenants and in PRD the list of tenants is stored in a Blob storage. By activating Redis features, where will this information be stored both in DEV and PRD?
3 - Assuming that list of tenants will be stored in a different store, Is there any migration we need to run to ensure that list of tenants is retrieved correctly? If yes, which are the steps?

Thanks

**Reply by Piedone** [reply]
Great!

1. I recommend using some Redis browser to quickly check if e.g. Data Protection key rings are saved there (those are the most long-living from the supported types of Redis-stored items).
2. Redis doesn't store tenant settings, so neither will change. However, for all cases I recommend https://docs.orchardcore.net/en/latest/reference/core/Shells/#database-shells-configuration-provider instead.
3. The list of tenants is not affected. Redis is only used for caching, not to store anything that should be durable.

**Reply by miguelisidoro** [reply]
@Piedone thanks.

I am using [RedisInsight](https://redis.com/redis-enterprise/redis-insight/) as a Redis browser.
I configured Redis settings as follows in my appsettings.json.

"OrchardCore_Redis": {
    "Configuration": "127.0.0.1:6379,allowAdmin=true", // Redis Configuration string.
    "InstancePrefix": "", // Optional prefix allowing a Redis instance to be shared by different applications.
    "DisableCertificateVerification": false // Disable SSL/TLS certificate verification.
}

When I try to connect to 127.0.0.1:6379 (I have deployed a Redis Docker container in my DEVBOX using port 6379 following https://redis.io/learn/operate/orchestration/docker), I get an error saying it can't connect.

![image](https://github.com/OrchardCMS/OrchardCore/assets/5429152/f6dfd6c0-e87b-4692-9b57-700767c659d3)

The RedisInsight interface is as follows:

![image](https://github.com/OrchardCMS/OrchardCore/assets/5429152/eaf5928a-4917-411d-a969-368cf642c2ab)

What am I missing here?

Thanks

**Reply by Piedone** [reply]
I can't tell from this, sorry. RedisInsight can't connect either, right? Because then that's some general Redis configuration issue.

**Reply by miguelisidoro** [reply]
@Piedone thanks. I am now able to connect to Redis using Redis Insight but I don't see Data Protection key rings or any other information. Once again I am missing something here...

1- In what moment are Data Protection key rings written to Redis?
2 - How do I know that Redis configuration in my web app is pointing correctly to my Redis deployment?
3 - Assuming a correct configuration in my web app, how should I check if I already have cached data in Redis?

Thanks

**Reply by Piedone** [reply]
All in all, if you have the Redis modules enabled, and you see no errors in the log after the app starts up, then Redis is working for you. Since it's only about caching, it's nothing readily visible, but if there are issues, you'll see it in the log.

**Reply by miguelisidoro** [reply]
@Piedone thanks. I dont see any errors in the log but I am still not confident that Redis is correctly configured. Just to summarize what I did:

1 - Configured Redis settings as follows in my appsettings.json.

"OrchardCore_Redis": {
"Configuration": "127.0.0.1:6379,allowAdmin=true", // Redis Configuration string.
"InstancePrefix": "", // Optional prefix allowing a Redis instance to be shared by different applications.
"DisableCertificateVerification": false // Disable SSL/TLS certificate verification.
}

2 - In Startup, activated the Redis features:

services
                .AddOrchardCms()
                .AddTenantFeatures("OrchardCore.Redis", "OrchardCore.Redis.Bus")
                .AddTenantFeatures("OrchardCore.Redis", "OrchardCore.Redis.Cache")
                .AddTenantFeatures("OrchardCore.Redis", "OrchardCore.Redis.Lock")

Is this correct and enough? If not, what is missing?

Thanks

**Reply by Piedone** [reply]
You can write `.AddTenantFeatures("OrchardCore.Redis", "OrchardCore.Redis.Bus", "OrchardCore.Redis.Cache", "OrchardCore.Redis.Lock")` but it's the same, just shorted. Otherwise yes, good.

**Reply by miguelisidoro** [reply]
@Piedone thanks.

Just to confirm, is there any other step I need to perform?

About using Redis in production, what are the settings I have to configure in appsetttings.json and in Azure? Is there any information you can share? We are thinking in using Azure Redis.

Thanks

**Reply by Piedone** [reply]
I don't think there's anything else. You need to configure the Azure connection string, but there shouldn't be anything else different.

**Reply by miguelisidoro** [reply]
@Piedone thanks. I just tried to create a new tenant just to try Redis and I am getting the following error in the logs:

2024-05-03 17:26:12.8772|OrchardCore.Redis.Startup|ERR|'Redis' features are not active on tenant 'Default' as the 'Configuration' string is missing or invalid: Value cannot be null. (Parameter 'configuration')

My configuration is on a file called appsetttings.development.json that is used in development mode of our solution as follows:

{
    "OrchardCore_Redis": {
        "Configuration": "127.0.0.1:6379,allowAdmin=true", // Redis Configuration string.
        "InstancePrefix": "", // Optional prefix allowing a Redis instance to be shared by different applications.
        "DisableCertificateVerification": false // Disable SSL/TLS certificate verification.
    }
}

Is there anything wrong with this configuration? I have a Docker container with a Redis image running at 127.0.0.1:6379 so the configuration string should be correct.

The validation of the configuration is done at https://github.com/OrchardCMS/OrchardCore/blob/dc0d4fc4f291a2eba383b4c3e21a6da3b647ab7d/src/OrchardCore.Modules/OrchardCore.Redis/Startup.cs

Thanks

**Reply by Piedone** [reply]
The error indicates that your configuration is not applied. I guess either your app is not running with environment settings called "development" (I think casing matters here, shouldn't it be "Development"?), or what you have in your appsettings is incorrect. Such configuration should be below an "OrchardCore" property, see https://github.com/OrchardCMS/OrchardCore/blob/dc0d4fc4f291a2eba383b4c3e21a6da3b647ab7d/src/OrchardCore.Cms.Web/appsettings.json

**Reply by miguelisidoro** [reply]
@Piedone thanks. In fact, after changing the configuration to be below an "OrchardCore" property, the error disappears.

After running the web application, I get the following in my Redis container.

![image](https://github.com/OrchardCMS/OrchardCore/assets/5429152/c5abcc4b-d782-4dad-9d08-198a15ac3c8f)

1 - Is this the expected content? If yes, does this mean that Redis is properly configured and the configuration is complete?
Our development tenant is called "Default", reason I assume that all keys in Redis Cache begin with "Default". Orchard Core prefixes all keys with the tenant name which allows multitenancy.

2 - Another question: where does Orchard Core caches shell settings? We activated Redis cache and there is nothing there and I assume that Orchard Core does not retrieve shell settings from blob storage or database in every request.

Thanks

**Reply by Piedone** [reply]
1. Looks good, yes.
2. It basically caches them [here](https://github.com/OrchardCMS/OrchardCore/blob/5a45ac58fc5aa24dfca1a1cf0e14b45690be158e/src/OrchardCore/OrchardCore/Shell/ShellSettingsManager.cs#L300), in a field.

**Reply by miguelisidoro** [reply]
@Piedone thanks. Is it possible for us to cache shell settings in Redis? Something that we can put on Startup that will cache shell settings in Redis? Caching shell settings in Redis would allow caching to be distributed which is important if we have more than one instance in our case in Azure. Without caching shell settings in Redis, and if we have more than one instance, if we for example create a new tenant,  the new tenant is not recognised in all instances.

**Reply by Piedone** [reply]
I think this is handled somehow; I might actually be wrong on why there are no repeated shell settings queries (a tenant, once running, will keep its `ShellSettings` object in memory until it's, even remotely via Redis, restarted, but there can be queries for fetching the shell configuration document for other tenants). But I'll need to leave the rest of this as an exercise for the reader :).
